pkg/ebpf/encoding.go

```diff
@@ -0,0 +1,180 @@
+	if strings.Contains(ContentTypeProtobuf, accept) {
+		return pSerializer
+	}
+
+	return jSerializer
+}
+
+// GetUnmarshaler returns the appropriate Unmarshaler based on the given content type
+func GetUnmarshaler(ctype string) Unmarshaler {
+	if strings.Contains(ContentTypeProtobuf, ctype) {
+		return pSerializer
+	}
+
+	return jSerializer
+}
+
+type protoSerializer struct{}
+
+func (p protoSerializer) Marshal(conns *Connections) ([]byte, error) {
+	agentConns := make([]*agent.Connection, len(conns.Conns))
+	for i, conn := range conns.Conns {
+		agentConns[i] = FormatConnection(conn)
+	}
+	payload := &agent.Connections{Conns: agentConns}
+	return proto.Marshal(payload)
+}
+
+func (p protoSerializer) Unmarshal(blob []byte) (*agent.Connections, error) {
+	conns := new(agent.Connections)
+	if err := proto.Unmarshal(blob, conns); err != nil {
+		return nil, err
+	}
+	return conns, nil
+}
+
+func (p protoSerializer) ContentType() string {
+	return ContentTypeProtobuf
+}
+
+type jsonSerializer struct {
+	marshaller jsonpb.Marshaler
+}
+
+func (j jsonSerializer) Marshal(conns *Connections) ([]byte, error) {
+	agentConns := make([]*agent.Connection, len(conns.Conns))
+	for i, conn := range conns.Conns {
+		agentConns[i] = FormatConnection(conn)
+	}
+	payload := &agent.Connections{Conns: agentConns}
+	writer := new(bytes.Buffer)
+	err := j.marshaller.Marshal(writer, payload)
+	return writer.Bytes(), err
+}
+
+func (j jsonSerializer) Unmarshal(blob []byte) (*agent.Connections, error) {
+	conns := new(agent.Connections)
+	reader := bytes.NewReader(blob)
+	if err := jsonpb.Unmarshal(reader, conns); err != nil {
+		return nil, err
+	}
+	return conns, nil
+}
+
+func (j jsonSerializer) ContentType() string {
+	return ContentTypeJSON
+}
+
+// FormatConnection converts a ConnectionStats into an agent.Connection
+func FormatConnection(conn ConnectionStats) *agent.Connection {
+	return &agent.Connection{
+		Pid:                int32(conn.Pid),
+		Laddr:              formatAddr(conn.Source, conn.SPort),
+		Raddr:              formatAddr(conn.Dest, conn.DPort),
+		Family:             formatFamily(conn.Family),
+		Type:               formatType(conn.Type),
+		TotalBytesSent:     conn.MonotonicSentBytes,
+		TotalBytesReceived: conn.MonotonicRecvBytes,
+		TotalRetransmits:   conn.MonotonicRetransmits,
+		LastBytesSent:      conn.LastSentBytes,
+		LastBytesReceived:  conn.LastRecvBytes,
+		LastRetransmits:    conn.LastRetransmits,
+		Direction:          agent.ConnectionDirection(conn.Direction),
+		NetNS:              conn.NetNS,
+		IpTranslation:      formatIPTranslation(conn.IPTranslation),
+	}
+}
+
+func formatAddr(addr util.Address, port uint16) *agent.Addr {
+	if addr == nil {
+		return nil
+	}
+
+	return &agent.Addr{Ip: addr.String(), Port: int32(port)}
+}
+
+func formatFamily(f ConnectionFamily) agent.ConnectionFamily {
+	switch f {
+	case AFINET:
+		return agent.ConnectionFamily_v4
+	case AFINET6:
+		return agent.ConnectionFamily_v6
+	default:
+		return -1
```

Comment:
Should we add a value `agent.ConnectionFamily_unknown` ?

Reply:
👍

Reply:
since this is mostly a cosmetic change I decided to leave as it is for the following reasons:
* The enum names have to be unique so I'd have to define `unknownType` and `unknownFamily`;
* The protobuf documentation recommends to avoid negative enums;
* I get a `The first enum value must be zero in proto3.` when adding it as the first entry